src/modules/navigation/NavigationState.js

```diff
@@ -14,70 +17,83 @@ export function switchTab(index) {
 }
 
 // Action creators
-export function pushRoute(state) {
-  return (dispatch, getState) => {
-    // conditionally execute push to avoid double
-    // navigations due to impatient users
-    if (!isNavigationAnimationInProgress(getState())) {
-      dispatch({type: PUSH_ROUTE, payload: state});
-    }
+export function pushRoute(route) {
+  return {
+    type: PUSH_ROUTE,
+    payload: route
   };
 }
 
 export function popRoute() {
   return {type: POP_ROUTE};
 }
 
-export function navigationCompleted() {
-  return {type: NAVIGATION_COMPLETED};
-}
-
-const initialState = fromJS(
-  createNavigationState('MainNavigation', 'App', [
-    createNavigationState('HomeTab', 'Home', [{key: 'Counter', title: 'Counter'}]),
-    createNavigationState('ProfileTab', 'Profile', [{key: 'Color', title: 'Color'}])
-  ]));
+// reducers for tabs and scenes are separate
+const initialState = fromJS({
+  tabs: {
+    index: 0,
+    routes: [
+      {key: 'HomeTab', title: 'HOME'},
+      {key: 'ProfileTab', title: 'PROFILE'}
+    ]
+  },
+  // Scenes for the `HomeTab` tab.
+  HomeTab: {
+    index: 0,
+    routes: [{key: 'Counter', title: 'Counter Screen'}]
+  },
+  // Scenes for the `ProfileTab` tab.
+  ProfileTab: {
+    index: 0,
+    routes: [{key: 'Color', title: 'Color Screen'}]
+  }
+});
 
 export default function NavigationReducer(state = initialState, action) {
   switch (action.type) {
-    case PUSH_ROUTE:
-      return state
-        .set('isNavigating', true)
-        .updateIn(['routes', state.get('index')], tabState =>
-          tabState
-            .update('routes', routes => routes.push(fromJS(action.payload)))
-            .set('index', tabState.get('routes').size));
-
-    case POP_ROUTE:
-      return state
-        .set('isNavigating', true)
-        .updateIn(['routes', state.get('index')], tabState =>
-          tabState
-            .update('routes', routes => routes.pop())
-            .set('index', tabState.get('routes').size - 2));
+    case PUSH_ROUTE: {
+      // Push a route into the scenes stack.
+      const route = action.payload;
+      const tabs = state.get('tabs');
+      const tabKey = tabs.getIn(['routes', tabs.get('index')]).get('key');
+      const scenes = state.get(tabKey).toJS();
+      let nextScenes;
+      // fixes issue #52
+      // the try/catch block prevents throwing an error when the route's key pushed
+      // was already present. This happens when the same route is pushed more than once.
+      try {
+        nextScenes = NavigationStateUtils.push(scenes, route);
+      } catch (e) {
+        nextScenes = scenes;
+      }
+      if (scenes !== nextScenes) {
+        return state.set(tabKey, fromJS(nextScenes));
+      }
+      return state;
+    }
```

Comment:
Using `NavigationStateUtils.push` which checks for suplicate `key` in the passed route